topic/machine-learning/llm-langchain/cratedb_rag_customer_support_langchain.ipynb

```diff
@@ -410,11 +460,160 @@
+      ]
+     },
+     "execution_count": 48,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
    "source": [
     "chat_completion.choices[0].message.content"
    ]
   },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## CrateDB &#129505; OpenSource\n",
+    "\n",
+    "In the second part of this notebook we build a Retrieval-Augmented Generation (RAG) system using CrateDB with Jina AI and the Llama open-source models. \n",
+    "\n",
+    "### Configure Jina AI API "
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Jina AI embeddings on Hugging Face are a suite of open-source models designed for encoding text data into high-dimensional vectors, which can be used in various natural language processing tasks such as semantic search, text classification, and clustering. These embeddings utilize the BERT architecture and are specially tuned for longer sequence lengths up to 8192 tokens, which is notably higher than the standard BERT model's capacity.\n",
+    "\n",
+    "To use Jina AI embeddings you need a Jina AI API key which can be easily obtained once you sign up for an account on the Jina AI website."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 42,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from langchain_community.embeddings import JinaEmbeddings\n",
+    "\n",
+    "getenvpass(\"JINA_API_KEY\", prompt=\"Jina API key:\")\n",
+    "embeddings = JinaEmbeddings()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Initialize new embedding store\n",
+    "\n",
+    "The next step initializes a vector search store in CrateDB using embeddings generated by Jina AI models and returns relevant documents using similarity search."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 50,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "store = CrateDBVectorSearch.from_documents(\n",
+    "    embedding=embeddings,\n",
+    "    documents=data,\n",
+    "    collection_name=COLLECTION_NAME,\n",
+    "    connection_string=CONNECTION_STRING,\n",
+    ")\n",
+    "documents = return_documents(store, my_question)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Connect to a local LLM "
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "One of the easiest ways to connect to a local LLM is to download one of the models from the Llamafile repository (https://github.com/Mozilla-Ocho/llamafile?tab=readme-ov-file) and run it locally. Upon initiating the llamafile, it not only sets up a web-based user interface for chatting at the local address http://127.0.0.1:8080/, but it also offers an endpoint for chat completions that is compatible with the OpenAI API.\n",
```

Comment:
At first, when just skimming into the content, it did not become clear to us quickly that this patch actually adds **two** more variants about how to interface with LLMs. One procedure uses the Jina AI API, where a remote LLM is interfaced with, similar to Open AI, and the other procedure uses llamafile for invoking the LLM on your own premises.

Sure enough, it is in the title already:

> using CrateDB with Jina AI and the Llama open-source models.

... but still it wasn't too clear from the very beginning that it is two distinct procedures you are illustrating here.

While it is clear that both new options fit into the "Open Source LLM" box well, we would like to slice their presentations differently, and in a flat manner, so we would actually have **three** dedicated ways how to interface with LLMs, presentation-wise:

- Open AI API (commercial, remote)
- Jina AI API (open-source, remote)
- llamafile (open-source, local)

In this spirit, when using separate notebooks, we would gain dedicated URLs to link to them, may be able to exclude specific ones on the CI/testing procedures selectively, for example when infrastructure or authentication tokens can not be provided there, and would also be able to extend that list of variants more easily in the future, instead of adding even more variants into the same notebook.

Reply:
Jina AI is used for calculating embeddings - we can alternatively use HuggingFace or another open-source algorithm. Here there are two procedures:

- Open AI Embedding + LLM models (commerical)
- Jina AI Embeddng + llamafile (open-source)